### ISSUE TITLE
Split Words into 2 fields instead of using RLC

### ISSUE THREAD

**Issue by ed255** [privacy-scaling-explorations/zkevm-specs#216]
Currently we're using RLC to compress words into a single value whenever we want do do lookups of words.
This design requires obtaining a randomness between all the circuits that will do Word RLC encoding (which imposes constraints on the structure of the aggregation circuit tree), and also requires refactoring circuits to separate the columns and configuration pre-RLC-Word and post-RLC-Word.
Also since the RLC uses a byte decomposition of the word, every time we recover the original value it will be 32 bytes, which we need to range check.  This is sometimes an overhead because we may recover the word to do arithmetic for which 16 bits limbs are used.

A different approach to doing the RLC of Words is to split a Word into 2 fields: the less significant 128 bits (`word_lo`) and the most significant 128 bits (`word_hi`).  With this change we don't need to get a shared randomness across all circuits that "share" Words (we still need shared randomness for other things).  We would still need to do some range check to guarantee that `word_lo` and `word_hi` are 128 bits.

This change doesn't imply removing the RLC, because we still need it for doing lookups of dynamic length arrays in a single lookup (mainly doing keccak lookups, where the input is a dynamic length array of bytes)

Please let me know if I missed any pro and con for Word-RLC and word split approaches!

The idea of splitting words into 2 fields instead of doing the RLC was (AFAIK) originally proposed by @han0110

**Comment by ed255** [comment]
Doing the lo/hi approach will require using 2 cells in the places where we were using a single cell to encode a Word.  For lookup-able tables that contain words (rw_table, tx_table, block_table, mpt_table) there are two approaches:

# 2 columns per Word
Every column that currently contains an RLC-encoded Word will become 2 columns with the lo and hi parts of the word

For example, in the tx_table the column `value` would become `valueLo` and `valueHi` where words would be split (for example, for the tag `GasPrice`).  

Cons:
- Each word column is doubled.  rw_table uses 4 word columns for `AccountStorage`, so we'd need 4 more columns there

# 2 rows per Word
Every row that currently contains one or more RLC-encoded Words will be split into 2 rows, where one has the lo part of the words and the other has the hi part of the words.  The tag used in these rows would need to be split into too: TagHi, TagLo.

For example, in the tx_table the tag `GasPrice` would become `GasPriceLo` and `GasPriceHi`.  In the `value` of `GasPriceLo` we'd find the lo part of the gasPrice, and in the `value` of `GasPriceHi` the hi part.

Cons:
- Each row with words is doubled
- We need extra constraints for lookups (lookup with word(s) becomes 2 lookups)

**Comment by DreamWuGit** [comment]
> We would still need to do some range check to guarantee that word_lo and word_hi are 128 bits.
if check 128 bits range by lookup , the table seems very big ?

**Comment by ed255** [comment]
> > We would still need to do some range check to guarantee that word_lo and word_hi are 128 bits.
> 
> if check 128 bits range by lookup , the table seems very big ?

Good point.  128 bit range lookup is not feasible.  So even though the words appear as two 128 bit values, we'd do range check on limbs (only on one circuit).  For example, for the rw_table we may do this check in the state circuit or the EVM circuit.  In the circuit where we do the range check we'll just allocate 8 x 16 bit limbs, which will be range checked via lookup table, and shown to correspond to the 128 bit value.

**Comment by ed255** [comment]
Update on this topic after a chat with @barryWhiteHat and @han0110:
- We're very confident that this change needs to be done eventually
- This change involves a big refactor because it touches most of the circuits
- The main pain point of the world RLC is getting a shared randomness between all the circuits.  This is "easy" to achieve if all the circuits are proved under a single aggregation circuit, and "difficult" if we need to build a tree of aggregation circuits.
- After a quick analysis, it seems we could fit all the current circuits under a single aggregation circuit (at least in an initial version)
- We conclude that doing the world split can be delayed until the [first milestone](https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/277) is completed

About how to approach the word split, I started looking into it and will leave it on hold for now.  Here's what I did:
- Figure out how the tables would look like: https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/277
- Here are some thoughts on design details for word split: https://hackmd.io/-4dSNwUZS6-06pjXajWA9A?both#Details-on-the-approach

**Comment by CPerezz** [comment]
I'm done with the review on https://hackmd.io/rF-ysYcKRzGO2jFrYeXDnA?both @ed255 

Happy to get feedback/replies on my comments. Anyway, so far the ideas are nice and I don't see many more ways to solve this.
